rfcs/require_webdriver_bidi.md

```diff
@@ -0,0 +1,121 @@
+
+The test runner can use the `Test`’s `require_webdriver_bidi` property to determine whether to enable BiDi support for the test. If the `require_webdriver_bidi` property is set to `true`, the test runner will create a session with BiDi support.
+
+Even though the `TestExecutor` needs information about whether to activate BiDi or not at `connect`, which happens before the runner receives the test to run, the configuration can be done at the point [the browser is configured](https://github.com/web-platform-tests/wpt/pull/48618/files#diff-0df24b5b583c460182e687f7dc7a6a79dd2cd3389bc4a96f48483f60fceb51f7R272). This data can be used later by the specific executor implementation to [decide which protocol to use](https://github.com/web-platform-tests/wpt/pull/48618/files#diff-c2f15328bc1ddfa8fb93c09ae41651e2bcfdad4f257932263a3e92c3f8deffceR248).
+
+#### Alternatives
+
+* [Add metadata to `.ini` files](#add-metadata-to-ini-files).
+
+#### Lint rules
+
+Some lint rules has to be added to verify is used correctly:
+* It is unique per test file.
+* It has the correct value (either `true` or `false`).
+
+### Deprecating of the `require_webdriver_bidi` metadata tag
+
+Expectations are that eventually all the implementations would support BiDi either via WebDriverBiDi, or via implementation-specific bindings. At that point, the `require_webdriver_bidi` metadata tag can be deprecated. This can be done by simply removing the tag.
+
+## Risks
+
+### Having metadata in the Test is not sufficient
+
+`TestExecutor` needs information about whether to activate BiDi or not at `connect`, which happens before the runner receives the test metadata. This can be addressed by passing the data to implementation-specific `TestharnessExecutor` via the implementation-specific `BrowserSettings` and  like in the [example](https://github.com/web-platform-tests/wpt/pull/48618/files#diff-0df24b5b583c460182e687f7dc7a6a79dd2cd3389bc4a96f48483f60fceb51f7R269).
+
+### Potential false-negative due to missing `require_webdriver_bidi` tag
+
+Some implementations can enable BiDi by default, while others would still need enabling it via `require_webdriver_bidi`. This could lead to false-negative test passes by the tests requiring BiDi but not having the `require_webdriver_bidi` tag. This risk can be mitigated by adding documentation describing the logic behind.
+
+## Alternatives considered
+
+### Add metadata to `.ini` files
+
+In this alternative approach, the `require_webdriver_bidi` metadata would be added to the corresponding `.ini` file. However, this approach was ultimately declined because the `.ini` file metadata is primarily concerned with the expectations of the test run rather than the specific API requirements. On the other hand, the `require_webdriver_bidi` metadata is directly related to the test API itself.
+
+### Extract testdriver BiDi API into a separate file
```

Comment:
A different alternative here would be to make `testdriver.js` have a `features` query parameter that it could check before enabling the features, and which could be supported in the manifest to extract the required features. So the author would need to write `<script src="/resources/testdriver.js?features=bidi"></script>`, and that would be used both by the js code to decide whether to enable the feature, and also the manifest code to set the flag on the test. The problem would be cases where you include `testdriver` in a support file but _not_ in the top-level test.

Reply:
To clarify: do you think it worse mentioning as analternative, or you insist we should implement this way instead?

Reply:
I thought of it when reviewing, and I think it's worth discussing here :)

The fact that it doesn't really work for tests that are using testdriver outside the test document seems like a notable downside because you'd still need some mechanism e.g. the meta tag in the test document. But it does naturally fail closed in the sense that if you don't specify the feature you can't use it, which is a notable advantage.

Reply:
I don't think think this approach would allow adding or not adding `bidi` in the `test_driver` object. So in fact in both cases of `testdriver.js?features=bidi` and `testdriver.js`, the `bidi` will be in the `test_driver` object. However, it's up to the backend to restrict the bidi calls if "bidi" parameter is not set. Please confirm I understood your proposal correctly, and if so, I think it makes sense and I will re-write the proposal accordingly.

Reply:
UPD: actually, `testdriver` can use `document.currentScript.src` to get it's query parameters, and enable/disable BiDi based on this.

Reply:
https://github.com/sadym-chromium/wpt_rfcs/blob/sadym/require_webdriver_bidi/rfcs/require_webdriver_bidi.md#add-feature-query-parameter-to-testdriverjs

Reply:
> UPD: actually, `testdriver` can use `document.currentScript.src` to get it's query parameters, and enable/disable BiDi based on this.

Note that this doesn’t work in `<script type=module>`. Is that a use case testdriver needs/wants to support?

Reply:
I started drafting this alternative in the https://github.com/web-platform-tests/rfcs/pull/214

Reply:
> Note that this doesn’t work in <script type=module>. Is that a use case testdriver needs/wants to support?

I think testdriver.is currently not a module but `import.meta.url` could be used if it becomes a module.